compiler/src/dotty/tools/dotc/core/TypeUtils.scala

```diff
@@ -189,5 +190,30 @@ class TypeUtils:
     def stripRefinement: Type = self match
       case self: RefinedOrRecType => self.parent.stripRefinement
       case seld => self
+
+    /** The constructors of this tyoe that that are applicable to `argTypes`, without needing
+     *  an implicit conversion.
+     *  @param adaptVarargs   if true, allow a constructor with just a varargs argument to
+     *                        match an empty argument list.
+     */
+    def applicableConstructors(argTypes: List[Type], adaptVarargs: Boolean)(using Context): List[Symbol] =
+      def isApplicable(constr: Symbol): Boolean =
+        def recur(ctpe: Type): Boolean = ctpe match
+          case ctpe: PolyType =>
+            if argTypes.isEmpty then recur(ctpe.resultType) // no need to know instances
+            else recur(ctpe.instantiate(self.argTypes))
+          case ctpe: MethodType =>
```

Comment:
It looks like a constructor with an implicit parameter will end up as a `MethodType(List(), List(), ImplicitMethodType(...))` here and it will crash later when typing.

So, an example like the following will still crash:
```scala
class MyFunctionWithImplicit(implicit args: String)

trait MyFunction0WithImplicit[+R] extends MyFunctionWithImplicit {
  def apply(): R
}

def fromFunction0[R](f: Function0[R]): MyFunction0WithImplicit[R] = () => f() // error
```

Reply:
Good catch, yes.